pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -0,0 +1,292 @@
+					})
+				}
+			}
+		}
+	}
+
+	e.endpointCache.Update(hostname, slice.Name, endpoints)
+
+	log.Infof("Handle EDS endpoint %s in namespace %s", svcName, slice.Namespace)
+
+	if features.EnableHeadlessService.Get() {
+		if obj, _, _ := c.services.informer.GetIndexer().GetByKey(kube.KeyFunc(svcName, slice.Namespace)); obj != nil {
+			svc := obj.(*v1.Service)
+			// if the service is headless service, trigger a full push.
+			if svc.Spec.ClusterIP == v1.ClusterIPNone {
+				c.xdsUpdater.ConfigUpdate(&model.PushRequest{Full: true, NamespacesUpdated: map[string]struct{}{slice.Namespace: {}}})
+				return
+			}
+		}
+	}
+
+	_ = c.xdsUpdater.EDSUpdate(c.clusterID, string(hostname), slice.Namespace, e.endpointCache.Get(hostname))
+}
+
+func (e endpointSliceController) AppendInstanceHandler(c *Controller) {
+	if e.cache.handler == nil {
+		return
+	}
+	e.cache.handler.Append(func(old, curr interface{}, event model.Event) error {
+		ep, ok := curr.(*discoveryv1alpha1.EndpointSlice)
+		if !ok {
+			tombstone, ok := curr.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				log.Errorf("Couldn't get object from tombstone %#v", curr)
+				return nil
+			}
+			ep, ok = tombstone.Obj.(*discoveryv1alpha1.EndpointSlice)
+			if !ok {
+				log.Errorf("Tombstone contained an object that is not an endpoints slice %#v", curr)
+				return nil
+			}
+		}
+
+		e.updateEDSSlice(c, ep, event)
+
+		return nil
+	})
+}
+
+func (e endpointSliceController) GetEndpointServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance {
+	endpointsForPodInSameNS := make([]*model.ServiceInstance, 0)
+	endpointsForPodInDifferentNS := make([]*model.ServiceInstance, 0)
+
+	for _, item := range e.cache.informer.GetStore().List() {
+		slice := *item.(*discoveryv1alpha1.EndpointSlice)
+		endpoints := &endpointsForPodInSameNS
+		if slice.Namespace != proxyNamespace {
+			endpoints = &endpointsForPodInDifferentNS
+		}
+
+		*endpoints = append(*endpoints, c.getProxyServiceInstancesByEndpointSlice(slice, proxy)...)
+	}
+
+	// Put the endpointsForPodInSameNS in front of endpointsForPodInDifferentNS so that Pilot will
+	// first use endpoints from endpointsForPodInSameNS. This makes sure if there are two endpoints
+	// referring to the same IP/port, the one in endpointsForPodInSameNS will be used. (The other one
+	// in endpointsForPodInDifferentNS will thus be rejected by Pilot).
+	return append(endpointsForPodInSameNS, endpointsForPodInDifferentNS...)
+}
+
+func (e *endpointSliceController) InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
+	labelsList labels.Collection) ([]*model.ServiceInstance, error) {
+	esLabelSelector := klabels.Set(map[string]string{discoveryv1alpha1.LabelServiceName: svc.Attributes.Name}).AsSelectorPreValidated()
+	var slices []*discoveryv1alpha1.EndpointSlice
+	err := cache.ListAllByNamespace(e.cache.informer.GetIndexer(), svc.Attributes.Namespace, esLabelSelector, func(i interface{}) {
```

Comment:
nit: `EndpointSliceLister.EndpointSlices(ns).List()` is  used more often.

Reply:
Nice, didn't know about that. I switched to that impl. For others reading, that method uses ListAllByNamespace internally